apps/emqx_dashboard/src/emqx_dashboard_monitor.erl

```diff
@@ -111,10 +110,14 @@ latest2time(Latest) -> erlang:system_time(millisecond) - (Latest * 1000).
 %%   ...
 %% ]
 %%
-granularity_adapter(List) when length(List) > 1000 ->
-    granularity_adapter(List, []);
 granularity_adapter(List) ->
-    List.
+    case length(List) > 1000 of
+        true ->
+            NewList = granularity_adapter(List, []),
+            granularity_adapter(NewList);
```

Comment:
Nit. I believe this will have somewhat poor runtime characteristics (especially with respect to GC pressure), as compared with, for example, downsampling M samples at a time, where M is `NSamples / 1000`. Granted though I don't have any idea how realistic it is to have more than 2 iterations.

Reply:
this algorithm seems to be simple enough, will make it faster in this iteration.
in the next iteration, i plan to downsample with time stamps aligned with wall-clock. i.e. will completely rewrite it.

Reply:
just can't wait, fixed it in the latest commit.